src/framework/recycledList/RecycledList.jsx

```diff
@@ -0,0 +1,220 @@
+
+      if (offsetScrollPosition < firstItemTopEdgePosition) {
+        // If we've scrolled the frist visible item just a tiny bit below the
+        // top edge of the grid, then we need to decrement the index.
+        let itemDistanceScrolled = 0;
+        while (
+          itemDistanceScrolled > distanceScrolled &&
+          firstItemIndex > 0
+        ) {
+          const item = items[firstItemIndex - 1];
+          itemDistanceScrolled -= this.props.itemHeightProvider(item);
+          firstItemIndex --;
+        }
+        firstItemOffset += itemDistanceScrolled;
+      }
+    }
+
+    return {
+      firstItemIndex,
+      firstItemOffset,
+    };
+  }
+
+  getLastItemState(items) {
+    const itemsCount = items.length;
+
+    // Calculate the index and offset of the last recycled item.
+    const recycledItemsCount = this.props.recycledItemsCount;
+
+    const lastItemIndex = Math.min(
+      this.state.firstItemIndex + recycledItemsCount,
+      itemsCount
+    );
+
+    let lastItemOffset = 0;
+    for (let i = lastItemIndex; i < itemsCount; i++) {
+      const item = items[i];
+      lastItemOffset += this.props.itemHeightProvider(item);
+    }
+
+    return {
+      lastItemIndex,
+      lastItemOffset,
+    };
+  }
+
+  createFakeItem(key, height) {
+    // We clone the element in case it's been customized via props already.
+    return React.cloneElement(this.props.fakeItemElement, {
+      key,
+      style: {
+        minHeight: height,
+      },
+    });
+  }
+
+  updateItemPositions(items = this.props.items) {
+    this.setState(Object.assign(
+      this.getFirstItemState(items),
+      this.getLastItemState(items)
+    ));
+  }
+
+  render() {
+    const items = this.props.items;
+    const itemsCount = items.length;
+
+    // Create recycled items.
+    const visibleItems = [
+      // Prepend the top fake item first.
+      this.createFakeItem('topFakeItem', this.state.firstItemOffset),
+    ];
+
+    for (
+      let index = this.state.firstItemIndex;
+      index <= this.state.lastItemIndex && index < itemsCount;
+      index++
+    ) {
+      visibleItems.push(items[index]);
+    }
+
+    // Append the bottom fake item last.
+    visibleItems.push(
+      this.createFakeItem('bottomFakeItem', this.state.lastItemOffset)
+    );
+
+    const rootElement = React.cloneElement(
+      this.props.rootElement,
+      {},
+      visibleItems
+    );
+
+    return rootElement;
+  }
+
+}
+
+RecycledList.propTypes = {
+  rootElement: React.PropTypes.element.isRequired,
+  fakeItemElement: React.PropTypes.element,
+  items: PropTypes.array.isRequired,
+  overflowDistance: PropTypes.number.isRequired,
+  recycledItemsCount: PropTypes.number.isRequired,
+  itemHeightProvider: PropTypes.func.isRequired,
+  // NOTE: Expects an instance of ScrollPosition.
+  scrollPosition: PropTypes.object.isRequired,
+};
+
+RecycledList.defaultProps = {
+  // NOTE: fakeItemElement needs to accept style as a prop.
+  fakeItemElement: <div />,
+};
```

Comment:
RecycledList receives a `scrollPosition` instance; it doesn't instantiate it itself. So since it doesn't own this instance, it seems like a bit of a weird side-effect for it to call `teardown`. But you bring up a good point that since it added a listener it should clean up after itself. I'll add a `removeListener` method which it can call in `componentWillUnmount`.